core/src/sumeragi/fault.rs

```diff
@@ -630,21 +280,1363 @@ pub fn run_sumeragi_main_loop<F>(
+                    == Role::ProxyTail
+                {
+                    let validating_peers = atomic_sumeragi_guard.current_topology.peers_set_a();
+                    let mut peer_signatures = vec![None; validating_peers.len()];
+                    let voting_block = voting_block_option.take().expect("take voting block");
+                    let voting_block_hash = voting_block.block.hash();
+                    loop {
+                        match incomming_message_receiver.try_recv() {
+                            Ok(msg) => match msg {
+                                Message::BlockSigned(block_signed) => {
+                                    let block = block_signed.block;
+
+                                    // I don't think we update the topology here. That ship has sailed.
+
+                                    let network_topology = &atomic_sumeragi_guard.current_topology;
+
+                                    let block_hash = block.hash();
+
+                                    if block_hash != voting_block_hash {
+                                        error!("Block hash does not match voting block hash.");
+                                    }
+
+                                    let valid_signatures = network_topology
+                                        .filter_signatures_by_roles(
+                                            &[Role::ValidatingPeer, Role::Leader],
+                                            block.verified_signatures(),
+                                        );
+
+                                    let mut number_of_votes = 0;
+                                    for i in 0..validating_peers.len() {
+                                        for sig in &valid_signatures {
+                                            if *sig.public_key() == validating_peers[i].public_key {
+                                                peer_signatures[i] = Some(sig.clone());
+                                            }
+                                        }
+
+                                        if peer_signatures[i].is_some() {
+                                            number_of_votes += 1;
+                                        }
+                                    }
+
+                                    info!(
+                                        %block_hash,
+                                        number_of_votes = number_of_votes,
+                                        required_number_of_votes = network_topology.min_votes_for_commit() - 1,
+                                        "Received a vote for block",
+                                    );
+
+                                    if number_of_votes < network_topology.min_votes_for_commit() - 1
+                                    {
+                                        continue;
+                                    }
+
+                                    let mut signatures = Vec::new();
```

Comment:
I think `filter_map` would be convinient here.

Reply:
It was indeed useful.